codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/JsonShapeDeserVisitor.java

```diff
@@ -180,8 +181,21 @@ protected void deserializeStructure(GenerationContext context, StructureShape sh
             });
         });
 
+        boolean isErrorStructure = shape.hasTrait(ErrorTrait.class);
+        if (isErrorStructure) {
+            writer.write("var errorMessage string");
+        }
+
         // Iterate through the decoder. The member visitor will handle popping tokens.
         writer.openBlock("for key, value := range shape {", "}", () -> {
+            if (isErrorStructure) {
+                writer.write("keyLower := strings.ToLower(key)");
```

Comment:
The cost of lowercasing every key in every structure is non-trivial, it's not something we want to add here. We should just check for `Message` and `message`.